tls/s2n_fingerprint.c

```diff
@@ -45,278 +34,142 @@ static S2N_RESULT s2n_assert_grease_value(uint16_t val)
+        uint32_t max_output_size, uint8_t *output, uint32_t *output_size, uint32_t *str_size)
 {
     POSIX_ENSURE(type == S2N_FINGERPRINT_JA3, S2N_ERR_INVALID_ARGUMENT);
-    POSIX_ENSURE(max_hash_size >= MD5_DIGEST_LENGTH, S2N_ERR_INSUFFICIENT_MEM_SIZE);
-    POSIX_ENSURE_REF(hash);
-    POSIX_ENSURE_REF(hash_size);
+    const struct s2n_fingerprint_method *method = &ja3_fingerprint;
+    POSIX_ENSURE(max_output_size >= method->hash_size, S2N_ERR_INSUFFICIENT_MEM_SIZE);
+
+    POSIX_ENSURE_REF(ch);
+    POSIX_ENSURE(!ch->sslv2, S2N_ERR_PROTOCOL_VERSION_UNSUPPORTED);
+    POSIX_ENSURE_REF(output);
+    POSIX_ENSURE_REF(output_size);
     POSIX_ENSURE_REF(str_size);
-    *hash_size = 0;
+    *output_size = 0;
     *str_size = 0;
 
-    /* The maximum size of the JA3 string is variable and could theoretically
-     * be extremely large. However, we don't need enough memory to hold the full
-     * string when calculating a hash. We can calculate and add the JA3 string
-     * to the hash in chunks, similarly to how the TLS transcript hash is
-     * calculated by adding handshake messages to the hash as they become
-     * available. After a chunk is added to the hash, the string buffer can be
-     * wiped and reused for the next chunk.
+    struct s2n_stuffer output_stuffer = { 0 };
+    POSIX_GUARD(s2n_blob_init(&output_stuffer.blob, output, max_output_size));
+
+    /* The maximum size of the hash input string is variable and could theoretically
+     * be extremely large. However, we don't need enough memory to hold the full string
+     * when calculating a hash. We can calculate and add the string to the hash in chunks,
+     * similarly to how the TLS transcript hash is calculated by adding handshake
+     * messages to the hash as they become available. After a chunk is added to the hash,
+     * the buffer can be wiped and reused for the next chunk.
      *
-     * The size of this buffer was chosen fairly arbitrarily.
-     */
-    uint8_t string_mem[50] = { 0 };
-    struct s2n_blob string_blob = { 0 };
-    struct s2n_stuffer string_stuffer = { 0 };
-    POSIX_GUARD(s2n_blob_init(&string_blob, string_mem, sizeof(string_mem)));
-    POSIX_GUARD(s2n_stuffer_init(&string_stuffer, &string_blob));
-
-    /* JA3 uses an MD5 hash.
-     * The hash doesn't have to be cryptographically secure,
-     * so the weakness of MD5 shouldn't be a problem.
+     * This ensures that our calculation requires a constant amount of memory.
+     *
+     * The size of the buffer is chosen to be the block size of most hashes.
      */
-    DEFER_CLEANUP(struct s2n_hash_state md5_hash = { 0 }, s2n_hash_free);
-    POSIX_GUARD(s2n_hash_new(&md5_hash));
-    if (s2n_is_in_fips_mode()) {
-        /* This hash is unrelated to TLS and does not affect FIPS */
-        POSIX_GUARD(s2n_hash_allow_md5_for_fips(&md5_hash));
-    }
-    POSIX_GUARD(s2n_hash_init(&md5_hash, S2N_HASH_MD5));
+    uint8_t hash_mem[64] = { 0 };
```

Comment:
Just curious: when a digest is being calculated, is there a reason to use the buffer at all? Like, couldn't you just immediately call hash update on every new string without having to worry about writing it to a buffer and flushing it when it's full?

Reply:
I was thinking that updating the hash in blocks would be more efficient, but testing suggests that it's not. Updating the hash seems significantly more efficient. Guess the extra copy was more expensive than I thought :/ I've switched to just writing directly to the hash.